source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -140,9 +141,26 @@ void AuthenticatorImpl::startVerify() {
   curr_token_ = std::move(tokens_.back());
   tokens_.pop_back();
 
-  jwt_ = std::make_unique<::google::jwt_verify::Jwt>();
-  ENVOY_LOG(debug, "{}: Parse Jwt {}", name(), curr_token_->token());
-  Status status = jwt_->parseFromString(curr_token_->token());
+  if (provider_) {
+    jwks_data_ = jwks_cache_.findByProvider(provider_.value());
+    if (jwks_data_->getJwtProvider().enable_token_cache()) {
+      jwt_ = jwks_data_->getTokenCache().find(curr_token_->token());
+      if (jwt_) {
+        use_cache_jwt_ = true;
+      }
+    }
+  }
+  Status status;
+  if (use_cache_jwt_) {
+    ENVOY_LOG(debug, "{}: Skip parsing, Jwt {} is in Cache.", name(), curr_token_->token());
+    status = Status::Ok;
+  } else {
+    ENVOY_LOG(debug, "{}: Parse Jwt {}", name(), curr_token_->token());
+    owned_jwt_ = std::make_unique<::google::jwt_verify::Jwt>();
+    status = owned_jwt_->parseFromString(curr_token_->token());
+    jwt_ = owned_jwt_.release();
```

Comment:
This is not good. It will have memory lea if this line is called, but line 293 is not called.

Reply:
Thanks!. I will call `owned_jwt_.release()`  in `insert `at line 293.

Reply:
Overall, I am very concern about this implementation.  It add too much cache special requirements to authentication code, making it too complicated, and error-prone.   The following knowledge should be hiding from authenticator

1) cache is enabled or not from config
2) cache is enabled or not for this request,  "provider" is valid
3) cache found item needs to be released
4) cache is taking ownership of jwt

Here is my thinking of re-design
create JwtCacheWrapper class,  it will be created for each token.

authenticator only call following 3 wraper functions:

std::unique_ptr<JwtCacheWrapper> jwt_cache_wapper =  createJwtCacheWrapper(provider, token);

// parse token, internally  will check cache.
status = wrapper->parseToken(jwt_)

status = wrapper->verifyToken(jwks)
internally it will:
if  cache_hit, cache release token
else { 
 verify token here
 if status is ok,  save to cache.
}

Reply:
class `JwtCacheWrapper` should be in Authenticator?

Reply:
I got another idea:
1) change JwtCache to 
*  JwtCache not to derided from SimpleLru,  instead embed it. 
*  handle enable_jwt_cache inside,  it false, it is no-op
*  take std::unique_ptr&& for Insert() function to take the ownership of owned_jwt_
* lookup function can use ScopedLookup, found item value can be released right away,


2) split function verifyKey() into verifyKey() and handleGoodJwt()

3) calling cache lookup at the current place, if found,  call handleGoodJwt()

4) caling cache insert at the current place, 

that is it, we only have two places calling jwt_cache functions.

Reply:
Done.